### PR TITLE
[MM-56585]Show only unreads filter stays active after changing the 'Group unreads' setting

### PR DESCRIPTION
**Pull request by apshada**
#### Summary

Only unreads channels are visible, with no categories, and the threads button may not be visible if there are no unread threads

#### Ticket Link
Fixes: [#26000](https://github.com/mattermost/mattermost/issues/26000)
[MM-56585](https://mattermost.atlassian.net/browse/MM-56585)

#### Checklist

- [ ] Added or updated unit tests (required for all new features)
- [ ] Has UI changes
- [ ] Includes text changes and localization file updates
- [ ] Have tested against the 5 core themes to ensure consistency between them.

#### Device Information
This PR was tested on: Pixel 6A, Android

#### Release Note

```release-note
NONE
```

[MM-56585]: https://mattermost.atlassian.net/browse/MM-56585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by larkox — CHANGES_REQUESTED** [review]
I am not a big fan of this solution, because we would hit the database every time we instantiate the SubHeader. Not sure how often will be that, but doesn't seem right.

I feel the solution should be somewhere around the categories component (`app/screens/home/channel_list/categories_list/categories/categories.tsx`) or other places where the `observeOnlyUnreads` is used.

What do you think?

**Comment by larkox** [comment]
@apshada Are you still working on this? Do you need any help?

**Comment by apshada** [comment]
> @apshada Are you still working on this? Do you need any help?

@larkox I am working on it, you can shortly see updated commits, if I need any help I will ask for it, 
Thanks

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 1]
```diff
@@ -49,17 +49,25 @@ const Categories = ({
...
+    const [showAllCategory, setShowAllCategory] = useState(onlyUnreads);
     const categoriesToShow = useMemo(() => {
+        const orderedCategories = [...categories];
+        orderedCategories.sort((a, b) => a.sortOrder - b.sortOrder);
+
+        if (onlyUnreads && unreadsOnTop) {
+            setShowAllCategory(false);
+        } else {
+            setShowAllCategory(onlyUnreads);
+        }
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:61]: Why calling a `setState` inside a useMemo?

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 2]
```diff
@@ -49,17 +49,25 @@ const Categories = ({
     const isTablet = useIsTablet();
     const switchingTeam = useTeamSwitch();
     const teamId = categories[0]?.teamId;
+    const [showAllCategory, setShowAllCategory] = useState(onlyUnreads);
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:52]: There is no need for state. You can get this information from the props already, and they are already causing the re-render. This can be simplified to something like...

```suggestion
    const showAllCategory = onlyUnreads && !unreadsOnTop;
```

File: app/components/threads_button/threads_button.tsx [hunk 3]
```diff
@@ -113,10 +111,6 @@ const ThreadsButton = ({
         return [container, icon, text, badge];
     }, [customStyles, isActive, onCenterBg, styles, unreads, isOnHome]);
 
-    if (groupUnreadsSeparately && (onlyUnreads && !isActive && !unreads && !mentions)) {
```
Comment (larkox) [inline app/components/threads_button/threads_button.tsx:116]: Why this change?

**Review by larkox — CHANGES_REQUESTED** [review]
A couple of questions, and a potential improvement.

File: app/components/threads_button/threads_button.tsx (hunk 3 above)
Comment (apshada) [reply app/components/threads_button/threads_button.tsx:116]: basically, this condition only works when `groupUnreadsSeparately` is true, and the expected behavior when `groupUnreadsSeparately` is true is we want the threads button to be always visible

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 2 above)
Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx:52]: @larkox This won't work when `onlyUnreads ` & `unreadsOnTop ` is `true`, if both of them are `true`, then we have to make `onlyUnreads` false, so it can render category  `Flatlist`

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 1 above)
Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx:61]: @larkox removing it from `useMemo`, will use something like
`
  useEffect(() => {
        if (onlyUnreads && unreadsOnTop) {
            setShowAllCategory(false);
        } else {
            setShowAllCategory(onlyUnreads);
        }
    }, [onlyUnreads, unreadsOnTop, showAllCategory])
`

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 2 above)
Comment (larkox) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx:52]: But if both `onlyUnreads` and `unreadsOnTop` are true, this new variable `showAllCategory` is false, and will do the trick, right?

Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx]: @larkox Got it, with this condition
`const showAllCategory = onlyUnreads && !unreadsOnTop;` we don't require any state, 
but one condition needs to be check if `onlyUnreads` and `unreadsOnTop` is true then we have to return `orderedCategories`, so it won't render empty state

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 4]
```diff
@@ -49,12 +49,19 @@ const Categories = ({
...
     const teamId = categories[0]?.teamId;
+    const showAllCategory = onlyUnreads && !unreadsOnTop;
+
     const categoriesToShow = useMemo(() => {
+        const orderedCategories = [...categories];
+        orderedCategories.sort((a, b) => a.sortOrder - b.sortOrder);
+
+        if (onlyUnreads && unreadsOnTop) {
+            return orderedCategories;
+        }
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:60]: Still not sure about this change.

If `unreadsOnTop` is set, it should take preference over `onlyUnreads`. So... we want the `Unreads` category and the rest of the categories, right?

So... if both are true, the next if (`onlyUnreads && !unreadsOnTop`) will skip, and the `unreadsOnTop` if will do the right thing (show unreads and the rest of categories).

Therefore... There is no need to any changes on this `useMemo`. Just the changes you have later in the render, right?

**Review by larkox — CHANGES_REQUESTED** [review]

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 4 above)
Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx:60]: @larkox  If `unreadsOnTop` is true, then it is rendering `<EmptyState>` with all the categories, because  `onlyUnreads ` is also true, that's why  `return orderedCategories;`

File: app/components/threads_button/threads_button.tsx [hunk 5]
```diff
@@ -53,9 +53,7 @@ type Props = {
 
 const ThreadsButton = ({
     currentChannelId,
-    groupUnreadsSeparately,
```
Comment (larkox) [inline app/components/threads_button/threads_button.tsx:56]: Remove this from the props type and the index file too.

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 6]
```diff
@@ -49,12 +49,19 @@ const Categories = ({
     const isTablet = useIsTablet();
     const switchingTeam = useTeamSwitch();
     const teamId = categories[0]?.teamId;
+    const showAllCategory = onlyUnreads && !unreadsOnTop;
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:52]: After doing some testing locally, this variable should probably called "showOnlyUnreadsCategory". With that name everything will be more clear.

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 7]
```diff
@@ -74,7 +81,7 @@ const Categories = ({
                     currentTeamId={teamId}
                     isTablet={isTablet}
                     onChannelSwitch={onChannelSwitch}
-                    onlyUnreads={onlyUnreads}
+                    onlyUnreads={showAllCategory}
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:84]: Only unreads only happen if `showOnlyUnreadsCategory`. Make sure you update the dependency of this callback.

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 8]
```diff
@@ -105,17 +112,17 @@ const Categories = ({
...
                 />
             </View>
             }
-            {!switchingTeam && !initiaLoad && !onlyUnreads && (
+            {!switchingTeam && !initiaLoad && !showAllCategory && (
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:125]: Now this should read as:
- If `showOnlyUnreadsCategory`, show the unread categories.
- Else, show a list with all the different categories.

**Review by larkox — CHANGES_REQUESTED** [review]
I tried this locally, and with the changes I propose it should be working fine. Let me know if you find anything I may not be seeing.

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 7 above)
Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx]: @larkox Updated

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 8 above)
Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx]: @larkox Updated

File: app/screens/home/channel_list/categories_list/categories/categories.tsx (hunk 6 above)
Comment (apshada) [reply app/screens/home/channel_list/categories_list/categories/categories.tsx]: @larkox Updated

File: app/components/threads_button/threads_button.tsx [hunk 9]
```diff
@@ -39,7 +39,6 @@ const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => ({
 
 type Props = {
     currentChannelId: string;
-    groupUnreadsSeparately: boolean;
     onCenterBg?: boolean;
     onlyUnreads: boolean;
```
Comment (larkox) [inline app/components/threads_button/threads_button.tsx:43]: Only unreads can be also removed. And therefore, the test that check this too.

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 10]
```diff
@@ -49,17 +49,20 @@ const Categories = ({
...
     const categoriesToShow = useMemo(() => {
-        if (onlyUnreads && !unreadsOnTop) {
-            return ['UNREADS' as const];
-        }
         const orderedCategories = [...categories];
         orderedCategories.sort((a, b) => a.sortOrder - b.sortOrder);
+
+        if (showOnlyUnreadsCategory) {
+            return ['UNREADS' as const];
+        }
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:60]: This is an uneeded change (the order change, I mean)

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 11]
```diff
@@ -89,7 +92,7 @@ const Categories = ({
                 />
             </>
         );
-    }, [teamId, intl.locale, isTablet, onChannelSwitch, onlyUnreads]);
+    }, [teamId, intl.locale, isTablet, onChannelSwitch, onlyUnreads, showOnlyUnreadsCategory]);
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:95]: `onlyUnreads` is no longer a dependency, isn't it?

**Review by larkox — CHANGES_REQUESTED** [review]
A few minor nitpicks.

File: app/components/threads_button/__snapshots__/threads_button.test.tsx.snap [hunk 12]
```diff
@@ -190,7 +190,98 @@ exports[`Thread item in the channel list Threads Component should match snapshot
 </View>
 `;
 
-exports[`Thread item in the channel list Threads Component should match snapshot with only unreads filter 1`] = `null`;
+exports[`Thread item in the channel list Threads Component should match snapshot with only unreads filter 1`] = `
```
Comment (larkox) [inline app/components/threads_button/__snapshots__/threads_button.test.tsx.snap:193]: The test that generates this snapshot probably makes no sense now, because the `onlyUnreads` no longer affects anything.

File: app/components/threads_button/index.ts [hunk 13]
```diff
@@ -2,30 +2,20 @@
...
     const currentTeamId = observeCurrentTeamId(database);
 
     return {
         currentChannelId: observeCurrentChannelId(database),
-        groupUnreadsSeparately: querySidebarPreferences(database, Preferences.CHANNEL_SIDEBAR_GROUP_UNREADS).
-            observeWithColumns(['value']).
-            pipe(
-                switchMap((prefs: PreferenceModel[]) => of$(getSidebarPreferenceAsBool(prefs, Preferences.CHANNEL_SIDEBAR_GROUP_UNREADS))),
-            ),
         onlyUnreads: observeOnlyUnreads(database),
```
Comment (larkox) [inline app/components/threads_button/index.ts:19]: You missed removing this one from here.

File: app/components/threads_button/threads_button.test.tsx [hunk 14]
```diff
@@ -29,7 +27,6 @@ describe('Thread item in the channel list', () => {
         const {toJSON} = renderWithIntlAndTheme(
```
Comment (larkox) [inline app/components/threads_button/threads_button.test.tsx:27]: Due to the changes, this test is the same as `Threads Component should match snapshot`, so can be removed.

File: app/components/threads_button/threads_button.test.tsx [hunk 15]
```diff
@@ -51,8 +48,6 @@ describe('Thread item in the channel list', () => {
...
```
Comment (larkox) [inline app/components/threads_button/threads_button.test.tsx:48]: Due to the changes, this test is the same as `Threads Component should match snapshot`, so can be removed.

File: app/screens/home/channel_list/categories_list/categories/categories.tsx [hunk 16]
```diff
@@ -49,17 +49,21 @@ const Categories = ({
...
+
         const orderedCategories = [...categories];
         orderedCategories.sort((a, b) => a.sortOrder - b.sortOrder);
+
         if (unreadsOnTop) {
             return ['UNREADS' as const, ...orderedCategories];
         }
         return orderedCategories;
-    }, [categories, onlyUnreads, unreadsOnTop]);
+    }, [categories, onlyUnreads, unreadsOnTop, showOnlyUnreadsCategory]);
```
Comment (larkox) [inline app/screens/home/channel_list/categories_list/categories/categories.tsx:66]: `onlyUnreads` is no longer a dependency, isn't it?

File: app/screens/home/channel_list/categories_list/subheader/subheader.tsx [hunk 17]
```diff
@@ -23,7 +23,6 @@ const style = StyleSheet.create({
 
 const SubHeader = ({unreadsOnTop}: Props) => {
     const showFilter = useSharedValue(!unreadsOnTop);
-
```
[the remark this answers is not in the document]
Comment (larkox) [reply app/screens/home/channel_list/categories_list/subheader/subheader.tsx:26]: There is still this unneeded change in this file.

**Review by larkox — CHANGES_REQUESTED** [review]
Some more clean up.

**Review by larkox — APPROVED** [review]
Looks good now. Thank you very much for all the changes!

**Comment by larkox** [comment]
@pvev This should be ready for your review.

**Review by pvev — APPROVED** [review]
LGTM! great work @apshada 🚀

**Review by yasserfaraazkhan — APPROVED** [review]
Screenshot 2024-04-16 at 10 27 26 AM
Screenshot 2024-04-16 at 10 27 39 AM
